core/src/main/java/io/dekorate/utils/TopologicalSort.java

```diff
@@ -0,0 +1,218 @@
+    for (Decorator item : items) {
+      Class identity = item.getClass();
+      Class[] after = item.after();
+      Class[] before = item.before();
+      Node node = getOrCreate(identity, unordered);
+
+      node.references.add(item);
+      if (after != null) {
+        // the after order is the same as the dependsOn order, so we simply add it.
+        for (Class a : after) {
+          List<Decorator> afterDecorators = dictionary.lookup(a);
+          if (afterDecorators != null) {
+            for (Decorator afterDecorator : afterDecorators) {
+              if (!afterDecorator.getClass().equals(identity)) {
+                node.depends.add(afterDecorator.getClass());
+              }
+            }
+          } else if (isVerbose()) {
+            LOGGER.info("[sort] Warning. Declared decorator in " + identity + " is not found " + a);
+          }
+        }
+      }
+
+      if (before != null) {
+        // for before order, we need to find the nodes that represent it and add the current node to its dependsOn field.
+        for (Class b : before) {
+          List<Decorator> beforeDecorators = dictionary.lookup(b);
+          if (beforeDecorators != null) {
+            for (Decorator beforeDecorator : beforeDecorators) {
+              if (!beforeDecorator.getClass().equals(identity)) {
+                Node beforeNode = getOrCreate(beforeDecorator.getClass(), unordered);
+                beforeNode.depends.add(identity);
+                // to avoid
+                node.depends.remove(beforeDecorator.getClass());
+              }
+            }
+          } else if (isVerbose()) {
+            LOGGER.info("[sort] Warning. Declared decorator in " + identity + " is not found " + b);
+          }
+        }
+      }
+    }
+    return unordered.values();
+  }
+
+  /**
+   * Sort nodes with dependencies using a Graph Topological algorithm plus the visitor pattern to avoid having cycles and
+   * to gain better performance.
+   */
+  private static List<Decorator> sortNodes(Collection<Node> items) {
+    List<Decorator> ordered = new ArrayList<>(items.size());
+    Set<Class> visited = new HashSet<>();
+    List<Node> cycle = new LinkedList<>();
+    // First, we loop over each node in order of insertion.
+    ArrayDeque<Node> back = new ArrayDeque<>();
+    back.addAll(items);
+
+    while (!back.isEmpty()) {
+      // Get the first node.
+      Node node = back.pop();
+      Class identity = node.value;
+      if (isVerbose()) {
+        LOGGER.info("[sort] Sort " + identity);
+      }
+      List<Class> dependants = node.depends;
+      if (dependants == null) {
+        cycle.clear();
+        visited.add(identity);
+        ordered.addAll(node.references);
+      } else {
+        boolean ready = true;
+        for (Class d : dependants) {
+          if (!visited.contains(d)) {
+            ready = false;
+            if (isVerbose()) {
+              LOGGER.info("[sort] " + identity + " is not ready because it needs " + d);
```

Comment:
Continuing my previous comment. This entry could be merged with the one above so that we have one line per node poped, or something like this.

Reply:
I keep this trace here because I deleted the previous one.